### PR TITLE
feat: cli: Add cli to inspect Eth addresses and simulate calls

### PR DESCRIPTION
**Pull request by shrenujbansal**
## Related Issues

https://github.com/filecoin-project/lotus/issues/9778

## Proposed Changes

> ➜  lotus git:(sbansal/eth-state-cli) ✗ ./lotus eth stat --filaddr t410fiux3jxq5bcdb3ctzgghvxt25ss3zgos2ifm3qki
> Filecoin address:  f410fiux3jxq5bcdb3ctzgghvxt25ss3zgos2ifm3qki
> Eth address:  0x452fb4de1d08861d8a79318f5bcf5d94b7933a5a
> Code cid:  bafk2bzacednmsvivt6tszxgnn3xy3iuwqqlhn5uif6mf4gj2du6ec4m6emble

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [x] New features have usage guidelines and / or documentation updates in
  - [x] [Lotus Documentation](https://lotus.filecoin.io)
  - [x] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [x] Tests exist for new functionality or change in behavior
- [x] CI is green

### CODE REVIEW

File: cli/eth.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+var EthCmd = &cli.Command{
+	Name:  "eth",
+	Usage: "Query eth contract state",
+	Subcommands: []*cli.Command{
+		EthGetAddressCmd,
+		EthCallSimulateCmd,
+	},
+}
+
+func ethAddrFromFilecoinAddress(ctx context.Context, addr address.Address, fnapi v0api.FullNode) (ethtypes.EthAddress, address.Address, error) {
```
Comment (arajasek) [inline cli/eth.go:27]: Let's move the helper to the end of the file?

File: cli/eth.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+var EthGetAddressCmd = &cli.Command{
+	Name:      "stat",
+	Usage:     "Print eth/filecoin addrs and code cid",
+	ArgsUsage: "",
+	Flags: []cli.Flag{
+		&cli.StringFlag{
+			Name:  "filaddr",
+			Value: "",
```
Comment (arajasek) [inline cli/eth.go:70]:
```suggestion
```

File: cli/eth.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+			Usage: "Filecoin address",
+		},
+		&cli.StringFlag{
+			Name:  "ethaddr",
+			Value: "",
```
Comment (arajasek) [inline cli/eth.go:75]:
```suggestion
```

File: cli/eth.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+	Usage:     "Print eth/filecoin addrs and code cid",
+	ArgsUsage: "",
+	Flags: []cli.Flag{
+		&cli.StringFlag{
+			Name:  "filaddr",
+			Value: "",
+			Usage: "Filecoin address",
+		},
+		&cli.StringFlag{
+			Name:  "ethaddr",
```
Comment (arajasek) [inline cli/eth.go:74]:
```suggestion
			Name:  "ethAddr",
```

File: cli/eth.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	return ethAddr, faddr, nil
+}
+
+var EthGetAddressCmd = &cli.Command{
+	Name:      "stat",
+	Usage:     "Print eth/filecoin addrs and code cid",
+	ArgsUsage: "",
+	Flags: []cli.Flag{
+		&cli.StringFlag{
+			Name:  "filaddr",
```
Comment (arajasek) [inline cli/eth.go:69]:
```suggestion
			Name:  "filAddr",
```

File: cli/eth.go [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+	ethAddr, err := ethtypes.EthAddressFromFilecoinAddress(faddr)
+	if err != nil {
+		return ethtypes.EthAddress{}, addr, err
+	}
+
+	return ethAddr, faddr, nil
+}
+
+var EthGetAddressCmd = &cli.Command{
+	Name:      "stat",
```
Comment (arajasek) [inline cli/eth.go:64]: Odd name for the command?

**Review by arajasek — COMMENTED** [review]
Thanks for this PR! It's really nice to see this stuff coming along.

I think there's a non-trivial product question around what  we want the "get address" method to return. At its simplest, we could do the following:

- given a filecoin address, only convert f410 addresses to Eth addresses (and fail everything else)
- given an eth address, simply invoke `ToFilecoinAddress`

We _could_ do more. We could convert any secp address into the corresponding f4 adderss in the EAM's namespace. We could be as aggressive as the `lookupEthAddress` method (see [here](https://github.com/filecoin-project/lotus/blob/d71f5a7d12ef56270c43c7837ff6b3f30f50b009/node/impl/full/eth.go#L723)).

I think we need to loop in the FVM's product team for design feedback here. In the meantime, I would propose implementing the simpler approach for filecoin -> eth and landing that:

- given a filecoin address, if f410 convert it
- if not f410, lookup the actor, lookup its `Address`, and if it's f410, convert it

File: cli/eth.go (hunk 6 above)
Comment (shrenujbansal) [reply cli/eth.go:64]: This was specified by steb. I don't have a strong opinion. Do you have a preferred name?

File: cli/eth.go [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+	Usage:     "Simulate an eth contract call",
+	ArgsUsage: "[from] [to] [params]",
+	Action: func(cctx *cli.Context) error {
+
+		var fromEthAddr ethtypes.EthAddress
+		fromAddr, err := hex.DecodeString(cctx.Args().Get(0))
+		if err != nil {
+			return err
+		}
+		copy(fromEthAddr[:], fromAddr)
```
Comment (geoff-vball) [inline cli/eth.go:106]: We should use EthAddressFromHex() here.

File: cli/eth.go [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+		var toEthAddr ethtypes.EthAddress
+		toAddr, err := hex.DecodeString(cctx.Args().Get(1))
+		if err != nil {
+			return err
+		}
+		copy(toEthAddr[:], toAddr)
```
Comment (geoff-vball) [inline cli/eth.go:113]: EthAddressFromHex() here as well

File: cli/eth.go (hunk 7 above)
Comment (geoff-vball) [reply cli/eth.go:106]: Better actually to use it on line 102. That way we only have 1 function call and don't have to pre-declare `fromEthAddr`

File: cli/eth.go [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+	ethAddr, err := ethtypes.EthAddressFromFilecoinAddress(faddr)
+	if err != nil {
+		return ethtypes.EthAddress{}, addr, err
+	}
+
+	return ethAddr, faddr, nil
+}
+
+var EthGetAddressCmd = &cli.Command{
+	Name:      "stat",
```
[the remark this answers is not in the document]
Comment (shrenujbansal) [reply cli/eth.go]: Could you refer to what the 2 cmd should be?
The idea of this cmd is "give me the fil/eth addr and ill give you all relevant info about this actor/contract"
I suppose I should change the name from EthGetAddressCmd to EthGetInfo

File: cli/eth.go (hunk 7 above)
Comment (shrenujbansal) [reply cli/eth.go:106]: sigh. I looked for a helper for a bit but couldn't find it. Thanks for pointing it out

File: cli/eth.go [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+			}
+		} else if ethAddr != "" {
+			if ethAddr[:2] == "0x" {
+				ethAddr = ethAddr[2:]
+			}
+			addr, err := hex.DecodeString(ethAddr)
+			if err != nil {
+				return err
+			}
+			copy(eaddr[:], addr)
```
Comment (geoff-vball) [inline cli/eth.go:72]: Oh, same thing with ethtypes.EthAddressFromHex() here

File: cli/eth.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (geoff-vball) [reply cli/eth.go]: `EthGetInfoFromFilecoinAddr` and `EthGetInfoFromEthAddr` maybe? I guess it's fine as is. Just seems like we're mostly doing two separate thing, even though the result comes from the same api call.

**Review by geoff-vball — APPROVED** [review]

**Pull request merged**